runtime/util/hshelp.c

```diff
@@ -482,6 +485,173 @@ areSingleSlotConstantRefsIdentical(J9ROMConstantPoolItem * romCP1, U_32 index1,
+	 */
+	J9ROMConstantPoolItem *romCP1 = J9_ROM_CP_FROM_ROM_CLASS(romClass1);
+	J9ROMConstantPoolItem *romCP2 = J9_ROM_CP_FROM_ROM_CLASS(romClass2);
+	J9SRP *callSiteData1 = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass1);
+	J9SRP *callSiteData2 = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass2);
+	/* get call site name and signature */
+	J9ROMNameAndSignature* nas1 = SRP_PTR_GET(callSiteData1 + callSiteIndex1, J9ROMNameAndSignature*);
+	J9ROMNameAndSignature* nas2 = SRP_PTR_GET(callSiteData2 + callSiteIndex2, J9ROMNameAndSignature*);
+	/* get call site BSM index */
+	U_16 *bsmIndices1 = (U_16 *) (callSiteData1 + romClass1->callSiteCount);
+	U_16 *bsmIndices2 = (U_16 *) (callSiteData2 + romClass2->callSiteCount);
+	U_16 bsmIndex1 = bsmIndices1[callSiteIndex1];
+	U_16 bsmIndex2 = bsmIndices2[callSiteIndex2];
+	/* get top of bsm data */
+	U_16 *bsmData1 = findBSMDataAtIndex(bsmIndices1 + romClass1->callSiteCount, bsmIndex1);
+	U_16 *bsmData2 = findBSMDataAtIndex(bsmIndices2 + romClass2->callSiteCount, bsmIndex2);
+	/* additional arg variables */
+	U_16 additionalArgCount1 = bsmData1[1];
+	U_16 additionalArgCount2 = bsmData2[1];
+	/* BSM MethodHandle reference */
+	J9ROMMethodHandleRef *mhRef1 = (J9ROMMethodHandleRef*) &romCP1[bsmData1[0]];
+	J9ROMMethodHandleRef *mhRef2 = (J9ROMMethodHandleRef*) &romCP2[bsmData2[0]];
+	U_32 bsmHandleTypeAndCpType1 = mhRef1->handleTypeAndCpType;
+	U_32 bsmHandleTypeAndCpType2 = mhRef2->handleTypeAndCpType;
+	U_32 bsmCpTypeIndex1 = bsmHandleTypeAndCpType1 & J9DescriptionCpTypeMask;
+	U_32 bsmCpTypeIndex2 = bsmHandleTypeAndCpType2 & J9DescriptionCpTypeMask;
+	UDATA bsmCpType1 = J9_CP_TYPE(J9ROMCLASS_CPSHAPEDESCRIPTION(romClass1), bsmCpTypeIndex1);
+	UDATA bsmCpType2 = J9_CP_TYPE(J9ROMCLASS_CPSHAPEDESCRIPTION(romClass2), bsmCpTypeIndex2);
+	/* additional helpers */
+	J9ROMMethodRef* bsmMethod = NULL;
+	J9ROMNameAndSignature *bsmNAS = NULL;
+	J9UTF8 *bsmSig = NULL;
+	U_8* sigData = NULL;
+	U_32 sigIndex = 0;
+	U_32 sigLength = -1;
+	U_16 i = 0;
+
+	/* compare method name and signature */
+	if (!areNameAndSigsIdentical(nas1, nas2)) {
+		return FALSE;
+	}
+
+	/* compare number of additional arguments */
+	if (additionalArgCount1 != additionalArgCount2) {
+		return FALSE;
+	}
+
+	/* Before continuing verify that the bsm MethodHandle refers to a Method and not a Field. The method
+	 * signature will be used to determine
+	 * whether the additional arguments are double or single slot which is impossible with a field.
+	 * The indy call will fail at a later point.
+	 */
+	if ((TRUE == isMethodHandleAField(bsmCpType1)) || (TRUE == isMethodHandleAField(bsmCpType2))) {
+		return FALSE;
+	}
+
+	/* verify that bsm MethodHandles are the same:
+	 * step 1) verify that the handle types and cp types match
+	 * step 2) verify that the method references are the same (at this point we know they are
+	 * both method references)
+	 */
+	if (bsmHandleTypeAndCpType1 != bsmHandleTypeAndCpType2) {
```

Comment:
If this is done before the `isMethodHandleAField()` call, then we only need to check if one of them is a field becasue we're sure they are the same type

Reply:
updated.